examples/smart-contract-rest-api.py

```diff
@@ -101,78 +82,108 @@ def custom_background_code():
 
 
-@app.route('/echo-post', methods=['POST'])
-@catch_exceptions
-@authenticated
-@json_response
-def echo_post(request):
+async def echo_post(request):
     # Parse POST JSON body
-    body = json.loads(request.content.read().decode("utf-8"))
+
+    body = await request.json()
 
     # Echo it
-    return {
+    res = {
         "post-body": body
     }
+    return web.json_response(data=res)
+
 
 #
-# Main method which starts everything up
+# Main setup method
 #
 
-
-def main():
+async def setup_and_start(loop):
     # Use TestNet
-    settings.setup_testnet()
+    settings.setup_privnet()
 
     # Setup the blockchain
     blockchain = LevelDBBlockchain(settings.chain_leveldb_path)
     Blockchain.RegisterBlockchain(blockchain)
-    dbloop = task.LoopingCall(Blockchain.Default().PersistBlocks)
-    dbloop.start(.1)
-    NodeLeader.Instance().Start()
+
+    p2p = NetworkService()
+    loop.create_task(p2p.start())
+    bg_task = loop.create_task(custom_background_code())
 
     # Disable smart contract events for external smart contracts
     settings.set_log_smart_contract_events(False)
 
-    # Start a thread with custom code
-    d = threading.Thread(target=custom_background_code)
-    d.setDaemon(True)  # daemonizing the thread will kill it when the main thread is quit
-    d.start()
-
-    # Hook up Klein API to Twisted reactor.
-    endpoint_description = "tcp:port=%s:interface=localhost" % API_PORT
+    app = web.Application()
+    app.add_routes([
+        web.route('*', '/', home_route),
+        web.get("/echo-get/{msg}", echo_msg),
+        web.post("/echo-post/", echo_post),
+    ])
 
-    # If you want to make this service externally available (not only at localhost),
-    # then remove the `interface=localhost` part:
-    # endpoint_description = "tcp:port=%s" % API_PORT
-
-    endpoint = endpoints.serverFromString(reactor, endpoint_description)
-    endpoint.listen(Site(app.resource()))
+    runner = web.AppRunner(app)
+    await runner.setup()
+    site = web.TCPSite(runner, "0.0.0.0", API_PORT)
+    await site.start()
 
     # Run all the things (blocking call)
     logger.info("Everything setup and running. Waiting for events...")
-    reactor.run()
-    logger.info("Shutting down.")
+    return site
+
+
+async def shutdown():
+    # cleanup any remaining tasks
+    for task in asyncio.Task.all_tasks():
+        with suppress(asyncio.CancelledError):
+            task.cancel()
+            await task
+
+
+def system_exit():
+    raise SystemExit
+
+
+def main():
+    loop = asyncio.get_event_loop()
+
+    # because a KeyboardInterrupt is so violent it can shutdown the DB in an unpredictable state.
+    loop.add_signal_handler(SIGINT, system_exit)
+
+    main_task = loop.create_task(setup_and_start(loop))
+
+    try:
+        loop.run_forever()
+    except SystemExit:
+        logger.info("Shutting down...")
+        site = main_task.result()
+        loop.run_until_complete(site.stop())
```

Comment:
This will result in 
```
AttributeError: 'NoneType' object has no attribute 'stop'
```

Reply:
Can you reproduce this? If so what are the steps? It doesn't seem to happen here

Reply:
I think I realized my mistake. I actually added this onto the code for `np-api-server` so that is probably why it didn't work. Sorry about that.